File: 프로그래머스/2/131130. 혼자 놀기의 달인/혼자 놀기의 달인.py
def recursive(cards, start, visited, li):
    visited[start] = True
    li.append(start)
    ns = cards[start] - 1

    if visited[ns]:
        return li

    return recursive(cards, ns, visited, li)


def solution(cards):
    visited = [False] * len(cards)
    result = []
    for i in range(len(cards)):
        if visited[i]:
            continue
        result.append(len(recursive(cards, i, visited, [])))

    result.sort(reverse=True)

    if len(result) >= 2:
        a, b = result[:2]
        return a * b
    else:
        return 0

File: 프로그래머스/2/131130. 혼자 놀기의 달인/혼자 놀기의 달인.java
import java.util.stream.*;
import java.util.*;

class Solution {
    List<Integer> scores = new ArrayList();
    public int solution(int[] cards) {
        boolean[] visited = new boolean[cards.length + 1];
        
        for(int i : cards){
            if (visited[i]) continue;
            int cnt = 0;
            int index = i;
            while (!visited[index]){
                cnt += 1;
                visited[index] = true;
                index = cards[index - 1];
            }
            
            scores.add(cnt);
        }
        
        if (scores.size() == 1) return 0;
        
        int[] arr = scores.stream()
            .sorted(Comparator.comparing((Integer i) -> -(int)i))
            .mapToInt(i->(int)i)
            .toArray();
        
        return arr[0] * arr[1];
    }
}